apps/src/aichat/redux/aichatRedux.ts

```diff
@@ -201,51 +198,85 @@ const saveAiCustomization = async (
     },
   };
 
+  // Notify the UI that a save is in progress.
+  dispatch(startSave(saveType));
+
   await Lab2Registry.getInstance()
     .getProjectManager()
     ?.save({source: JSON.stringify(trimmedCurrentAiCustomizations)}, true);
+};
 
-  dispatch(setSavedAiCustomizations(trimmedCurrentAiCustomizations));
-
-  const changedProperties = findChangedProperties(
-    savedAiCustomizations,
-    trimmedCurrentAiCustomizations
-  );
-  if (
-    changedProperties.some(property =>
-      [
-        'selectedModelId',
-        'temperature',
-        'systemPrompt',
-        'retrievalContexts',
-      ].includes(property)
-    )
-  ) {
-    dispatch(setNewChatSession());
-  }
+// Thunk called after a save has completed successfully.
+// Updates the chat window and reports analytics as necessary.
+export const onSaveComplete =
+  () => (dispatch: AppDispatch, getState: () => RootState) => {
+    const {savedAiCustomizations, currentAiCustomizations, currentSaveType} =
+      getState().aichat;
 
-  changedProperties.forEach(property => {
-    dispatch(
-      addChatMessage({
-        id: getNewMessageId(),
-        role: Role.MODEL_UPDATE,
-        chatMessageText:
-          AI_CUSTOMIZATIONS_LABELS[property as keyof AiCustomizations],
-        status: Status.OK,
-        timestamp: getCurrentTime(),
-      })
+    const changedProperties = findChangedProperties(
+      savedAiCustomizations,
+      currentAiCustomizations
     );
-    if (eventDescription) {
-      analyticsReporter.sendEvent(
-        eventDescription,
-        {
-          propertyUpdated: property,
-          levelPath: window.location.pathname,
-        },
-        PLATFORMS.BOTH
+    if (
+      changedProperties.some(property =>
+        [
+          'selectedModelId',
+          'temperature',
+          'systemPrompt',
+          'retrievalContexts',
+        ].includes(property)
+      )
+    ) {
+      dispatch(setNewChatSession());
+    }
+
+    changedProperties.forEach(property => {
+      dispatch(
+        addChatMessage({
+          id: getNewMessageId(),
+          role: Role.MODEL_UPDATE,
+          chatMessageText:
+            AI_CUSTOMIZATIONS_LABELS[property as keyof AiCustomizations],
+          status: Status.OK,
+          timestamp: getCurrentTime(),
+        })
       );
+
+      if (currentSaveType) {
+        analyticsReporter.sendEvent(
+          saveTypeToAnalyticsEvent[currentSaveType],
+          {
+            propertyUpdated: property,
+            levelPath: window.location.pathname,
+          },
+          PLATFORMS.BOTH
+        );
+      }
+    });
+
+    // Update our last saved customizations to match the current customizations
+    dispatch(setSavedAiCustomizations(currentAiCustomizations));
```

Comment:
Not sure if it matters, but is there a reason not to do this at the top of this function? It seems like maybe? the most important save callback, but is done after a bunch of other stuff.

Reply:
Yeah, not a strong reason either way, I think my thinking at the time was to update redux as the last step after all other actions were complete, but any time after the save completes works.